### PR TITLE
upgrade api version

### PR DESCRIPTION
**Pull request by benhinchley**
I accidentally closed this PR https://github.com/sajari/sdk-js/pull/164 (tbh not sure how I did this still)

This is just a continuation of that PR, with a reduced scope to just be the api version upgrade.

### CODE REVIEW

**Review by wwalser — APPROVED** [review]

**Comment by wwalser** [comment]
Looking forward to integrating it!

File: src/client.ts [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+  protected async trackEvent(req: TrackEventRequest) {
+    hasNetworkConnection();
+
+    const request = this.createRequest("trackEvent");
+    const resp = await fetch(request, {
+      method: "POST",
+      body: JSON.stringify(req),
+    });
+
+    if (resp.status !== 200) {
```
[the remark this answers is not in the document]
Comment (wwalser) [reply src/client.ts:121]: Is ok a short hand specifically for 200 or for anything in the 200 range?

File: src/tracking.ts [hunk 2]
```diff
@@ -115,48 +108,11 @@ export class SearchIOAnalytics {
...
-        new Error(message)
-      );
-    }
-
-    return response.json();
+    return this.trackEvent({
+      type,
+      query_id: queryId,
+      ...data,
+    } as unknown as TrackEventRequest);
```
[the remark this answers is not in the document]
Comment (wwalser) [reply src/tracking.ts:115]: Yes.
Not particularly dark either, gets used all over the place when the programmer is confident in the type of something but TS can't infer the correct thing.

File: src/index.ts [hunk 3]
```diff
@@ -393,30 +162,26 @@ class QueryPipeline extends EventEmitter {
           ...(neuralScore && { neuralScore }),
           ...(featureScore && { featureScore }),
           score,
-          values: processProtoValues(values),
-          token: t,
```
[the remark this answers is not in the document]
Comment (wwalser) [reply src/index.ts:397]: Are we removing deprecated things given the major version bump?

File: src/client.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (JasonBerry) [reply src/client.ts:121]: > The `ok` read-only property of the [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response) interface contains a Boolean stating whether the response was successful (status in the range 200-299) or not.

File: src/tracking.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (JasonBerry) [reply src/tracking.ts:115]: Wouldn't just `as TrackEventRequest` suffice for that?

[the remark this answers is not in the document]
Comment (wwalser) [reply src/tracking.ts:115]: Usually coercing directly to the type doesn't work for reasons that I knew at some time but have forgotten.

[the remark this answers is not in the document]
Comment (benhinchley) [reply src/tracking.ts:115]: `...data,` spreads to the correct type, but Typescript can't infer that correctly. So one has to basically say the type is `unknown` and then retype it as `as TrackEventRequest`. It is a particularly shady practice.

File: src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (benhinchley) [reply src/index.ts:397]: I had initially thought of removing all the deprecated tracking stuff (as this was a major version bump), but happy to remove that part of the change if we aren't open to this part of the change.

File: src/index.ts [hunk 4]
```diff
@@ -84,202 +35,36 @@ export class Client {
    * const client = new Client("<account_id>", "<collection_id>", "<endpoint>");
    * ```
    *
-   * @param project
+   * @param account
    * @param collection
    * @param {string} [endpoint]
    */
   constructor(
```
[the remark this answers is not in the document]
Comment (benhinchley) [reply src/index.ts:42]: Ahh interesting, will give this a go.

File: src/client.ts [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+    }
+  }
+
+  private createRequest(verb: "query" | "trackEvent") {
+    const requestInit = {
+      method: "POST",
+      headers: new Headers({
+        "Account-Id": this.account,
+        Accept: "application/json",
+        "Content-Type": "text/plain",
```
[the remark this answers is not in the document]
Comment (benhinchley) [reply src/client.ts:57]: This is a hack, and there was a comment on this code at some point (I will go find it), but by setting the `content-type` to `text/plain` we remove the need for a `OPTIONS` request to be sent first.

**Review by wwalser — APPROVED** [review]

File: src/client.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (JasonBerry) [reply src/client.ts:47]: Gross. I remember that now.

File: src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (JasonBerry) [reply src/index.ts:397]: Major version, kill it.

File: test/index-server.test.ts [hunk 6]
```diff
@@ -107,16 +106,18 @@ describe("Client", () => {
     );
   });
 
-  test("works with non-JSON erroring response", async () => {
+  // NOTE(bhinchley): not sure this test is needed, not sure why the api
+  // would return a non json response.
+  test.skip("works with non-JSON erroring response", async () => {
```
Comment (JasonBerry) [inline test/index-server.test.ts:111]: I think you've handled this so we can un-skip it now?

In general I've seen non-JSON responses come back when load balancers fall over and things of that ilk.

**Review by JasonBerry — APPROVED** [review]
Good to go, just had one comment on one of the tests we'd skipped.

**Pull request merged**